extensions/s3/src/main/java/io/deephaven/extensions/s3/S3AsyncClientFactory.java

```diff
@@ -0,0 +1,144 @@
+                        // If we find that the STANDARD retry policy does not work well in all situations, we might
+                        // try experimenting with ADAPTIVE retry policy, potentially with fast fail.
+                        // .retryPolicy(RetryPolicy.builder(RetryMode.ADAPTIVE).fastFailRateLimiting(true).build())
+                        .retryPolicy(RetryMode.STANDARD)
+                        .apiCallAttemptTimeout(instructions.readTimeout().dividedBy(3))
+                        .apiCallTimeout(instructions.readTimeout())
+                        // Adding a metrics publisher may be useful for debugging, but it's very verbose.
+                        // .addMetricPublisher(LoggingMetricPublisher.create(Level.INFO, Format.PRETTY))
+                        .scheduledExecutorService(ensureScheduledExecutor())
+                        .build())
+                .region(Region.of(instructions.regionName()))
+                .credentialsProvider(instructions.awsV2CredentialsProvider());
+        instructions.endpointOverride().ifPresent(builder::endpointOverride);
+        final S3AsyncClient ret = builder.build();
+        if (log.isDebugEnabled()) {
+            log.debug().append("Creating S3AsyncClient for region: ").append(instructions.regionName()).endl();
+        }
+        return ret;
+    }
+
+    private static class HttpClientConfig {
+        private final int maxConcurrentRequests;
+        private final Duration connectionTimeout;
+
+        HttpClientConfig(final int maxConcurrentRequests, final Duration connectionTimeout) {
+            this.maxConcurrentRequests = maxConcurrentRequests;
+            this.connectionTimeout = connectionTimeout;
+        }
+
+        int maxConcurrentRequests() {
+            return maxConcurrentRequests;
+        }
+
+        Duration connectionTimeout() {
+            return connectionTimeout;
+        }
+
+        @Override
+        public int hashCode() {
+            int result = maxConcurrentRequests;
+            result = 31 * result + connectionTimeout.hashCode();
+            return result;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            final HttpClientConfig that = (HttpClientConfig) o;
+            return maxConcurrentRequests == that.maxConcurrentRequests
+                    && connectionTimeout.equals(that.connectionTimeout);
+        }
+    }
+
+    private static SdkAsyncHttpClient getOrBuildHttpClient(@NotNull final S3Instructions instructions) {
+        final HttpClientConfig config = new HttpClientConfig(instructions.maxConcurrentRequests(),
+                instructions.connectionTimeout());
+        return httpClientCache.computeIfAbsent(config, key -> AwsCrtAsyncHttpClient.builder()
```

Comment:
I worry a little bit about whether client sharing will lead to interference. These objects are `AutoCloseable`.